web/src/chartOptions/dataset.ts

```diff
@@ -0,0 +1,62 @@
+import {ECFeature, MaxSegmentLength, SegmentCoords, WgsCovPlotDB} from "../db";
+import {constant, isNil, sum, times, values} from "lodash";
+import {getFluGeneFeature, getMaxSegmentLength, getSegmentCoords} from "./segmented/getSegmentsInfo";
+import {setState, state} from "../state";
+import {leaveSelect} from "echarts/types/src/util/states";
+
+export const getDatasets = (db: WgsCovPlotDB) => {
+  let datasets = [];
+  if (!isNil(db.segments)) { // segmented virus
+    /* For nf-flu, setup must be done before setting chart option:
+    Set up gene feature, position array and segment coords
+    */
+    let maxSegmentLength: MaxSegmentLength = getMaxSegmentLength(state);
+    setState({"maxSegmentLength": maxSegmentLength}) // disable shallow merging and fully replace
```

Comment:
@nhhaidee Avoid using setState in this function and others that are meant to be without side-effects, simply returning an updated object value. State updates should be done elsewhere and carefully with the use of "untrack"/"unwrap" to ensure that multiple value updates to the state store only create one update of the chart. If you're not careful, multiple chart updates will be queued and multiple diffs of the store nodes will be performed - absolutely killing performance and negating the benefits of using SolidJS in the first place.

See https://github.com/peterk87/wgscovplot/blob/b01f861c6e1e5e252c446cc59714e4ce42c4166e/web/src/ui/ChartContainer.tsx#L247